### ISSUE TITLE
worldgen or oregen issues - tin ores and possibly more do not spawn correctly

### ISSUE THREAD

**Issue by DrTasslehoff** [FTBTeam/FTB-Interactions#968]
## Modpack Version: 
2.0.2

## Bug Report:
certain seeds seem to not generate everything properly and I'm not sure why. It's been a small thing people have mentioned since the pack has released but after actually getting a seed and checking it personally it's very obvious that something is weird, and it's not strictly something with pack configs or something. 

Seed with expected ore spawns: 3641920093901290765 
You can see plenty of tin and stibnite veins without much effort. 

Two seeds without tin and stibnite veins:
7162897440616446793 
-1799545939522390749 
the top seed is something I found myself, and the second is one from someone else on the discord. 

On the second seed, -1799545939522390749 , coords (46, 106,-242) is around the middle of an area that is a good example of an ore not existing that should

**Comment by Saereth** [comment]
really no idea how to even begin to address this

**Comment by talchas** [comment]
TL;DR - ocean at 0,0 seeds are fucked, because GTCE screwed up. They used a BlockPos they intended to be chunk-relative with world methods that are not chunk-relative to determine the height of ground in the area. There's a whole bunch of other screwups which would make the height limits more subtly broken, but they don't matter because everything just looks around 0,0 instead.


More stream-of-debugging writeup of how I got there:

The 7162897440616446793 is over the ocean though? The ground in the ocean is less than tin's min_height, and thus it doesn't generate tin. (And because :gtce:, it rolls tin before checking if tin is valid for the height)

For -1799545939522390749 at 46 -242, that is gtce grid 1, -5, and it generated a rock sphere for that coordinate and a couple nearby ones:

1 -5: overworld/interactions_black_granite_sphere.json success 69,25,-207
1 -4: overworld/interactions_copper_vein.json success 86,13,-150
2 -5: overworld/interactions_magnetite_vein.json success 136,23,-199
2 -4: overworld/interactions_tin_vein.json failed 40 >= 0 (40-120/0)
0 -5: overworld/interactions_basalt_sphere.json success 22,20,-225
0 -4: overworld/interactions_iron_vein.json success 43,11,-155
1 -6: overworld/interactions_red_granite_sphere.json success 56,27,-287
2 -6: overworld/Strata Exclusive/interactions_apatite_swamp.json success 135,21,-247
0 -6: overworld/interactions_marble_sphere.json success 21,11,-267

so there was a cancelled tin vein in 96,-192 to 144,-144 because for some reason gtce thought the max height around there was 0. gtce plz. When I ran it again having my character start here rather than fly in this direction or `forge gen 0 0 0 1000 100` it was 32 instead.

Max height is determined when the first chunk that is generated touches the grid entry and looks at the closest chunk corner to the grid center and takes the getTopSolidOrLiquidBlock of that. This would be dumb enough because it would matter what direction you approached the grid from, but it gets worse - when gtce is asked to generate a chunk, it sets up the nine surrounding _grids_ using that chunk so that it can ask them if any ore spilled over into this chunk. So the height it is measuring can be a hundred blocks away in both x/z from the block it would choose as the center of the vein. And indeed dumping more info we get:

2 -4 (100): coming from 3 -7
2 -4 (100): master init 63/32 from 15,0,0
2 -4: overworld/interactions_tin_vein.json failed 40 >= 32 (40-120/32 vs 12)

So it looks at the height of 63,-112 to decide if the tin should generate, and I guess that's the ocean or something? Looking in world yes, but the height should be more like 48 (which would still be a problem), and that's assuming that the name is a typical minecraft lie and water doesn't count. I might be off by one in my math somewhere I guess.

Deleting the world and running
forge generate 100 0 -190 10 100

2 -4 (100): coming from 6 -12
2 -4: overworld/interactions_tin_vein.json failed 40 >= 32 (40-120/32 vs 12)

and we have a cached master entry because we're coming from generating a chunk in this grid, so it's saved a height from generating this chunk from a different grid here (which is actually vaguely correct, though it's some random corner). Regardless it's still getting 63 getHeight and 32 or 33 getTopSolidOrLiquidBlock for some reason. 

Wait

I assumed GTCE wasn't screwing up. getHeight/getTopSolidOrLiquidBlock aren't chunk-relative, so GTCE is always asking about the world origin for heights.

**Comment by Saereth** [comment]
for cross reference https://github.com/GregTechCE/GregTech/issues/1152
